test/ARCHITECTURE.md

```diff
@@ -0,0 +1,257 @@
+will also be added to the affected test list 
+
+After having collected the destructive, nondestructive and affected tests a
+scheduling loop is started, if a machine was provided it is used for the
+nondestructive tests, destructive tests will always spawn a new machine. If no
+machine is provided a pool of global machines is created based on the provided
+`--jobs` and nondestructive tests. The test runner will first try to assign all
+nondestructive tests on the available global machines and start the tests. 
+
+A test is started by the `Test` class by calling the `start()` method executed
+the provided `command` (e.g. `./test/verify/check-apps --machine 127.0.0.1:2201
+--browser 127.0.0.1:9091`) with a `timeout` to cancel hanging tests after a
+timeout automatically and creates a temporarily file to store the output of
+the test `command` in. Finally the test is added to the `running_tests` list.
+
+The test runner inspects all `running_tests` in a tight loop and polls if the
+test is still running. If the process stopped or exited the test output and
+`returncode` is saved. Depending on the `returncode` the test runner makes a
+decision on what to do next as described in the diagram below.
+```mermaid
+graph TD;
+    finished[Test finished]
+    succeeded{Test succeeded?}
+    affected{"Test affected?"}
+    affecteddone["Retry three times"]
+    skipped{Test skipped?}   
+    todosuccess{Test todo?}
+    todonosuccess{Test todo?}
+    todosucceeded["Unexpected success
+    show as failure"]
+    todofail[Expected failure]
+    pixel_journal_failure{"Pixel test or
+    unexpected journal
+    message?"}
+    retry["Retry three times
+    to be robust against
+    test failures"]
+    testfailed["Test failed"]
+    failure_policy{"Known issue?"}
+    
+    finished --> skipped
+    skipped --> |No| succeeded
+    skipped --> |Yes| skiptest[Show as skipped test]
+    succeeded --> |No| todonosuccess
+    succeeded --> |yes| affected
+   
+    affected --> |Yes| affecteddone
+    affected --> |No| todosuccess
+    
+    todosuccess --> |Yes| todosucceeded
+    todosuccess --> |No| done[Test succeeded]
+    
+    todonosuccess --> |Yes| todofail
+    todonosuccess --> |No| failure_policy
+    
+    failure_policy --> |Yes| known_issue[Show as known issue]
+    failure_policy --> |No| pixel_journal_failure
+    
+    pixel_journal_failure --> |Yes| testfailed
+    pixel_journal_failure --> |No| retry
+```
+
+* `Skipped` -  tests can be skipped because the test can not run on the given `TEST_OS`.
+* `Affected` - tests to be retried to make sure any changes affecting them do not lead to flaky tests.
+* `Todo` - tests which are incomplete and expected to fail.
+* `Known issue` - Naughties are expected test failures due to known issues in software we test,
+                  we still run the test but if the test error output matches a known naughty it
+                  is skipped. The [bots](github.com/cockpit-project/bots) repository keeps track
```

Comment:
`https://` before "github.com/cockpit-project/bots", the link doesn't work otherwise.